habitat_baselines/il/data/data.py

```diff
@@ -0,0 +1,254 @@
+import os
+from typing import Callable, List, Tuple
+
+import cv2
+import numpy as np
+import torch
+import webdataset as wds
+import webdataset.filters as filters
+from tqdm import tqdm
+
+import habitat
+from habitat import logger
+from habitat.core.simulator import ShortestPathPoint
+from habitat.datasets.utils import VocabDict
+from habitat_baselines.utils.common import (
+    base_plus_ext,
+    create_tar_archive,
+    delete_folder,
+    get_scene_episode_dict,
+    valid_sample,
+)
+
+
+class EQADataset(wds.Dataset):
+    """Pytorch dataset for Embodied Q&A (both VQA and PACMAN)"""
+
+    def __init__(
+        self,
+        config,
+        input_type,
+        num_frames=5,
+        max_controller_actions=5,
+    ):
+        """
+        Args:
+            config: Config
+            input_type (string): Type of model being trained ("vqa", "pacman")
+            num_frames (int): number of frames used as input to VQA model
+            max_controller_actions (int):
+        """
+        self.config = config.TASK_CONFIG
+        self.input_type = input_type
+        self.num_frames = num_frames
+
+        with habitat.Env(config=self.config) as self.env:
+            self.episodes = self.env._dataset.episodes
+
+            # sorting and making episode ids consecutive for simpler indexing
+            self.sort_episodes()
+
+            self.q_vocab = self.env._dataset.question_vocab
+            self.ans_vocab = self.env._dataset.answer_vocab
+
+            self.eval_save_results = config.EVAL_SAVE_RESULTS
+
+            if self.config.DATASET.SPLIT == config.EVAL.SPLIT:
+                self.mode = "val"
+            else:
+                self.mode = "train"
+
+            self.frame_dataset_path = config.FRAME_DATASET_PATH.format(
+                split=self.mode
+            )
+
+            # [TODO] can be done in mp3d_eqa_dataset when loading
+            self.calc_max_length()
+            self.restructure_ans_vocab()
+
+            group_by_keys = filters.Curried(self.group_by_keys_)
+            super().__init__(
+                urls=self.frame_dataset_path + ".tar",
+                initial_pipeline=[group_by_keys()],
+            )
```

Comment:
Hard to understand what's happening here. Better self descriptive variables needed or comments.